src/transaction/log_manager.c

```diff
@@ -9788,3 +9788,121 @@ log_set_db_restore_time (THREAD_ENTRY * thread_p, INT64 db_restore_time)
 
   LOG_CS_EXIT (thread_p);
 }
+
+
+/*
+ * log_get_undo_record () - gets undo record from log lsa adress
+ *   return: S_SUCCESS or ER_code
+ * 
+ * thread_p (in):
+ * lsa_addr (in):
+ * page (in):
+ * record (in/out):
+ */
+SCAN_CODE
+log_get_undo_record (THREAD_ENTRY * thread_p, LOG_PAGE * log_page_p, LOG_LSA process_lsa, RECDES * recdes)
+{
+  LOG_RECORD_HEADER *log_rec_header = NULL;
+  struct log_mvcc_undo *mvcc_undo = NULL;
+  struct log_mvcc_undoredo *mvcc_undoredo = NULL;
+  int udata_length;
+  int udata_size;
+  char *undo_data;
+  LOG_LSA oldest_prior_lsa;
+  bool is_zipped = false;
+  char log_buf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
+
+  /* assert log record is not in prior list */
+  oldest_prior_lsa = *log_get_append_lsa ();
+  assert (LSA_LT (&process_lsa, log_get_append_lsa ()));
+
+  log_rec_header = LOG_GET_LOG_RECORD_HEADER (log_page_p, &process_lsa);
+  LOG_READ_ADD_ALIGN (thread_p, sizeof (*log_rec_header), &process_lsa, log_page_p);
+
+  if (log_rec_header->type == LOG_MVCC_UNDO_DATA)
+    {
+      LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*mvcc_undo), &process_lsa, log_page_p);
+      mvcc_undo = (struct log_mvcc_undo *) (log_page_p->area + process_lsa.offset);
+
+      udata_length = mvcc_undo->undo.length;
+      LOG_READ_ADD_ALIGN (thread_p, sizeof (*mvcc_undo), &process_lsa, log_page_p);
+    }
+  else if (log_rec_header->type == LOG_MVCC_UNDOREDO_DATA)
+    {
+      LOG_READ_ADVANCE_WHEN_DOESNT_FIT (thread_p, sizeof (*mvcc_undoredo), &process_lsa, log_page_p);
+      mvcc_undoredo = (struct log_mvcc_undoredo *) (log_page_p->area + process_lsa.offset);
+
+      udata_length = mvcc_undoredo->undoredo.ulength;
+      LOG_READ_ADD_ALIGN (thread_p, sizeof (*mvcc_undoredo), &process_lsa, log_page_p);
+    }
+
+
+  /* get undo record */
+  if (ZIP_CHECK (udata_length))
+    {
+      /* Get real size */
+      udata_size = (int) GET_ZIP_LEN (udata_length);
+      is_zipped = true;
+    }
+  else
+    {
+      udata_size = udata_length;
+    }
+
+  /* 
+   * If data is contained in only one buffer, pass pointer directly.
+   * Otherwise, allocate a contiguous area, copy the data and pass this area.
+   * At the end deallocate the area.
+   */
+  if (process_lsa.offset + udata_size < (int) LOGAREA_SIZE)
+    {
+      undo_data = (char *) log_page_p->area + process_lsa.offset;
+    }
+  else
+    {
+      /* Need to copy the data into a contiguous area */
+      char *area = PTR_ALIGN (log_buf, MAX_ALIGNMENT);
+
+      /* consider size of undo record is not larger than IO_MAX_PAGE_SIZE */
+      assert (udata_size <= IO_MAX_PAGE_SIZE);
```

Comment:
The records processed in log_undo_record () shouldn't be overflow records because of the different way of logging REC_BIGONE updates. In this way, IO_PAGESIZE is the maximum size one can have.